src/commands/BootstrapCommand.js

```diff
@@ -100,150 +121,139 @@ export default class BootstrapCommand extends Command {
+    async.parallelLimit(actions, this.concurrency, (err) => {
+      this.progressBar.terminate();
+      callback(err);
     });
   }
 
-  linkBinariesForPackage(pkg, callback) {
-    const actions = this.packages
-      .filter((dep) => this.hasMatchingDependency(pkg, dep) && dep.bin)
-      .map((dep) => (cb) => this.createBinaryLink(pkg, dep, cb));
-
-    async.parallelLimit(actions, this.concurrency, callback);
-  }
-
-  createBinaryLink(pkg, dep, callback) {
-    const dest = path.join(pkg.nodeModulesLocation, ".bin");
-
-    // The `bin` in a package.json may be either a string or an object.
-    // Normalize to an object.
-    const bins = typeof dep.bin === "string"
-      ? { [dep.name]: dep.bin }
-      : dep.bin;
-
-    async.series([(cb) => FileSystemUtilities.mkdirp(dest, cb)].concat(
-      Object.keys(bins).map((name) => (cb) => FileSystemUtilities.symlink(
-        path.join(dep.location, bins[name]),
-        path.join(dest, name),
-        cb
-      ))
-    ), callback);
-  }
-
-  installExternalPackages(pkg, callback) {
-    const allDependencies = pkg.allDependencies;
-
-    const externalPackages = Object.keys(allDependencies)
-      .filter((dependency) => {
-        const match = find(this.packages, (pkg) => {
-          return pkg.name === dependency;
+  /**
+   * Symlink all packages to the packages/node_modules directory
+   * Symlink package binaries to dependent packages' node_modules/.bin directory
+   * @param {Function} callback
+   */
+  symlinkPackages(callback) {
+    this.logger.info("Symlinking packages and binaries");
+    this.progressBar.init(this.filteredPackages.length);
+    const actions = [];
+    this.filteredPackages.forEach((filteredPackage) => {
+      const packageLocation = path.join(this.repository.packagesLocation, filteredPackage.name);
+      // actions to run for this package
+      const packageActions = [];
+      Object.keys(filteredPackage.allDependencies)
+        // filter out external dependencies and incompatible packages
+        .filter((dependency) => {
+          const match = find(this.packages, { name: dependency});
+          return match && filteredPackage.hasMatchingDependency(match);
+        })
+        .forEach((dependency) => {
+          // get path to dependency and its package.json
+          const dependencyLocation = path.join(this.repository.packagesLocation, dependency);
+          const dependencyPackageJsonLocation = path.join(dependencyLocation, "package.json");
+          // ignore dependencies without a package.json file
+          if (!FileSystemUtilities.existsSync(dependencyPackageJsonLocation)) {
+            this.logger.error(`Unable to find package.json for ${dependency} dependency. Skipping...`);
+          } else {
+            const installedDepLocation = path.join(filteredPackage.nodeModulesLocation, dependency);
+            // check if dependency is already installed
+            if (FileSystemUtilities.existsSync(installedDepLocation)) {
```

Comment:
Hi @rygine - Just tried this out on my primary React Server repo (been testing on a fresh clone until now).

Ran into some unexpected errors:

![image](https://cloud.githubusercontent.com/assets/115908/18178038/9a2ccdfa-7031-11e6-84c7-835ef45eeeae.png)

I think there a few things going on here:

- The `isDepSymlink !== dependencyLocation` check is true when `isDepSymlink` is false.
- I _think_ this whole block should be looking at `dependencyLinkLocation` (defined below) instead of `installedDepLocation`.  In other words it looks like we're intending to check what exists in the location where we want to put a symlink, but are instead checking what exists at the actual package location...